docs/proposals/20210722-openyurt-convertor-operator.md

```diff
@@ -0,0 +1,175 @@
+---
+title: OpenYurt Convertor Operator
+authors:
+  - "@gnunu"
+  - "@lindayu17"
+reviewers:
+  - "@rambohe-ch"
+  - "@guofei"
+  - "@kadisi"
+  - "@wenjun93"
+creation-date: 2021-07-22
+last-updated: 2021-07-22
+status: provisional
+---
+
+# Convert Vanilla Kubernetes cluster to OpenYurt cluster through Operator
+
+## Table of Contents
+* [Convert Vanilla Kubernetes cluster to OpenYurt cluster through Opertaor](#convert-vanilla-kubernetes-cluster-to-openyurt-cluster-through-operator)
+   * [Table of Contents](#table-of-contents)
+   * [Glossary](#glossary)
+   * [Summary](#summary)
+   * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-Goals/Future Work](#non-goalsfuture-work)
+   * [Proposal](#proposal)
+      * [User Stories](#user-stories)
+         * [Story 1](#story-1)
+         * [Story 2](#story-2)
+      * [Requirements](#requirements)
+         * [Functional Requirements](#functional-requirements)
+            * [FR1](#fr1)
+            * [FR2](#fr2)
+            * [FR3](#fr3)
+         * [Non-Functional Requirements](#non-functional-requirements)
+            * [NFR1](#nfr1)
+            * [NFR2](#nfr2)
+      * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Risks and Mitigations](#risks-and-mitigations)
+   * [Alternatives](#alternatives)
+   * [Upgrade Strategy](#upgrade-strategy)
+   * [Additional Details](#additional-details)
+      * [Test Plan](#test-plan)
+   * [Implementation History](#implementation-history)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+This OpenYurt Convertor Operator is to translate a vanilla Kubernetes cluster into an OpenYurt cluster, through a simple API (YurtConvetor CRD).
+
+## Motivation
+
+To make OpenYurt translation easy for user, i.e., as automative as possible.
+
+### Goals
+- Convert nodes automatically
+- Revert nodes automatically
+- Translated OpenYurt cluster should be the same as through Yurtcutl or doing manually
+
+### Non-Goals/Future Work
+
+This is NOT a replacement of yurtctl command.
+
+## Proposal
+
+This is an operator, with CRD and Controller.
+
+CRD's definition:
+
+// YurtconvertorSpec defines the desired state of Yurtconvertor
+type YurtconvertorSpec struct {
```

Comment:
The "YurtConvertor" does not sound like a CRD name for declaring a "state". How about "YurtCluster"?

The CRD SPEC should include CloudNodes[] and EdgeNodes[].  `Convert` and `Revert` are actions, not states. The controller should figure out the nodes that need to be reverted, hence we don't need the Revert array in the Spec.

To describe the Cloud/Edge Node lists, we can try the following struct

```
type YurtClusterSpec struct{
       CloudNodes ItemList
       EdgeNodes ItemList

       TunnelService    ComponentConfig    
       AppManager  ComponentConfig
}

// You may consider different config for Tunnel and AppManager separately to allow customization for each component
type ComponentConfig{
    Enable bool 
    Version string
}


type ItemList {
     //options
     Names []string
     //optional
     Pattern string // For regular expression
     //optional
     ExcludedNames []string
     //optional
     ExcludedPattern string 
} 


```